datafusion/core/src/config.rs

```diff
@@ -17,616 +17,517 @@
-        let configs = Self::new();
-        let mut docs = "| key | type | default | description |\n".to_string();
-        docs += "|-----|------|---------|-------------|\n";
 
-        let config_definitions: Vec<_> = configs
-            .config_definitions
-            .into_iter()
-            .map(normalize_for_display)
-            .collect();
+        let mut s = Self::default();
+        s.built_in.execution.target_partitions = 0; // Normalize for display
 
-        for config in config_definitions.iter().sorted_by_key(|c| c.key.as_str()) {
+        let mut docs = "| key | default | description |\n".to_string();
+        docs += "|-----|---------|-------------|\n";
+        for entry in s.entries() {
             let _ = writeln!(
                 &mut docs,
-                "| {} | {} | {} | {} |",
-                config.key, config.data_type, config.default_value, config.description
+                "| {} | {} | {} |",
+                entry.key,
+                entry.value.as_deref().unwrap_or(""),
+                entry.description
             );
         }
         docs
     }
 }
 
-/// Normalizes a config definition prior to markdown display
-fn normalize_for_display(mut v: ConfigDefinition) -> ConfigDefinition {
-    // Since the default value of target_partitions depends on the number of cores,
-    // set the default value to 0 in the docs.
-    if v.key == OPT_TARGET_PARTITIONS {
-        v.default_value = ScalarValue::UInt64(Some(0))
-    }
-    v
+/// [`ConfigExtension`] provides a mechanism to store third-party configuration within DataFusion
+///
+/// Unfortunately associated constants are not currently object-safe, and so this
+/// extends the object-safe [`ExtensionOptions`]
+pub trait ConfigExtension: ExtensionOptions {
+    const PREFIX: &'static str;
 }
 
-/// Configuration options struct. This can contain values for built-in and custom options
-#[derive(Clone)]
-pub struct ConfigOptions {
-    options: HashMap<String, ScalarValue>,
+/// An object-safe API for storing arbitrary configuration
+pub trait ExtensionOptions: Send + Sync + std::fmt::Debug + 'static {
+    fn as_any(&self) -> &dyn Any;
+
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+
+    fn cloned(&self) -> Box<dyn ExtensionOptions>;
+
+    fn set(&mut self, key: &str, value: &str) -> Result<()>;
+
+    fn entries(&self) -> Vec<ConfigEntry>;
 }
 
-/// Print the configurations in an ordered way so that we can directly compare the equality of two ConfigOptions by their debug strings
-impl Debug for ConfigOptions {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("ConfigOptions")
-            .field(
-                "options",
-                &format!("{:?}", BTreeMap::from_iter(self.options.iter())),
-            )
-            .finish()
+/// A type-safe container for [`ConfigExtension`]
+#[derive(Debug, Default, Clone)]
+pub struct Extensions(BTreeMap<&'static str, ExtensionBox>);
```

Comment:
This formulation is specifically designed to be able to replace the AnyMap currently on SessionConfig. As an added bonus it no longer requires types to be interior-mutable